### PR TITLE
fix(cmp): discover plugins relative to app path (#13940)

### LINKED ISSUE #13940: v2.7.4 CMP regression: the plugin isn't called
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Starting v2.7.4 my CMP plugin is no longer being called.

**To Reproduce**

I have the following config configuration:

```yaml
apiVersion: v1
data:
  nix.conf: |
    build-users-group = nixbld
    sandbox = false
    experimental-features = nix-command flakes
    substituters = https://cache.nixos.org https://nixhelm.cachix.org
    trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= nixhelm.cachix.org-1:esqauAsR4opRF0UsGrA6H3gD21OrzMnBBYvJXeddjtY=
  passwd: |
    nix:x:999:30000:Nix build user 1:/home/nix:/bin/false
    root:x:0:0::/root:/bin/bash
  plugin.yaml: '{"apiVersion":"argoproj.io/v1alpha1","kind":"ConfigManagementPlugin","metadata":{"name":"nix-cmp-plugin"},"spec":{"discover":{"fileName":"flake.nix"},"generate":{"command":["sh","-c","cat
    result"]},"init":{"command":["sh","-c","export OUTPUT=\"${ARGOCD_ENV_NIX_OUTPUT:-kubernetesConfiguration}\"\necho
    -ne \"Building for $OUTPUT\\n\\n\"\nif [ \"$PARAM_VALUES\" != \"\" ]; then\n  echo
    \"$PARAM_VALUES\" > values.json\n  nix-shell -p git --run ''git add values.json''\nfi\nnix
    build \".#${OUTPUT}\"\n"]},"name":"nix-cmp-plugin","version":"v1.0"}}'
kind: ConfigMap
metadata:
  labels:
    argocd.argoproj.io/instance: cloudops-argocd
  name: nix-cmp-config
  namespace: argocd
```

(pretty-printed that is:)

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ConfigManagementPlugin
metadata:
  name: nix-cmp-plugin
spec:
  discover:
    fileName: flake.nix
  generate:
    command:
      - sh
      - '-c'
      - cat result
  init:
    command:
      - sh
      - '-c'
      - |
        export OUTPUT="${ARGOCD_ENV_NIX_OUTPUT:-kubernetesConfiguration}"
        echo -ne "Building for $OUTPUT\n\n"
        if [ "$PARAM_VALUES" != "" ]; then
          echo "$PARAM_VALUES" > values.json
          nix-shell -p git --run ''git add values.json''
        fi
        nix build ".#${OUTPUT}"
  name: nix-cmp-plugin
  version: v1.0
```

Starting from 2.7.4 the plugin doesn't seem to be called at all (I tried to add some verbose logging into `init.command` to no avail, also it hard refreshes too fast). The pipeline runs without errors but suggests all the resources should be removed.

**Expected behavior**

Expected the CMP work just as it was working in 2.7.3. Reverting argo to 2.7.3 makes it operational again.

**Version**

2.7.4 installed via helm.

**Logs**

No useful logs in repo server.

### PR DESCRIPTION
**Pull request by crenshaw-dev**
According to our docs, CMP discovery was always meant to be relative to the repo root. That was a reasonable design, because discovery for other config management tools (e.g. Helm) is done at the repo root.

But those other tools have multiple outputs: the names of the directories which they can handle. CMP discovery only has a boolean output: "yes, I support that directory" or "no, I do not support that directory."

So, the actual implementation (in contradiction to the docs) answers that question at the app's path instead of the repo root. That's reasonable, because Argo CD needs to know "will this plugin work in _this directory_?" Since we only get a yes/no, we have to ask the question about a specific directory, instead of about the whole repo.

https://github.com/argoproj/argo-cd/pull/13360 made one important fix: instead of just sending the app directory to the plugin, we send the whole repo. That makes sense, because stuff outside the app directory might be important.

But the PR made a mistake: it changed the discovery mechanism to evaluate discovery rules relative to the repo root instead of the app path.

This PR does two things:
1) It returns CMP discovery to be relative to the app path instead of repo root.
2) It corrects the docs.

This PR does not revert the changes from https://github.com/argoproj/argo-cd/pull/13360 which makes the whole repo available to the plugin.

Fixes #13940

### CODE REVIEW

File: cmpserver/plugin/plugin.go [hunk 1]
```diff
@@ -315,12 +316,20 @@ func (s *Service) matchRepositoryGeneric(stream MatchRepositoryStream) error {
...
 
-func (s *Service) matchRepository(ctx context.Context, workdir string, envEntries []*apiclient.EnvEntry) (isSupported bool, isDiscoveryEnabled bool, err error) {
+func (s *Service) matchRepository(ctx context.Context, workdir string, envEntries []*apiclient.EnvEntry, appRelPath string) (isSupported bool, isDiscoveryEnabled bool, err error) {
 	config := s.initConstants.PluginConfig
 
+	appPath, err := securejoin.SecureJoin(workdir, appRelPath)
+	if err != nil {
+		log.WithFields(map[string]interface{}{
+			common.SecurityField:    common.SecurityHigh,
+			common.SecurityCWEField: 459,
```
Comment (leoluz) [inline cmpserver/plugin/plugin.go:326]: What `459` means in this case? Maybe better to add that code in a constant with an intuitive name?

Comment (crenshaw-dev) [reply cmpserver/plugin/plugin.go:326]: Good idea, I'll add a constant.

Comment (crenshaw-dev) [reply cmpserver/plugin/plugin.go]: Fixed!

**Review by leoluz — APPROVED** [review]
LGTM

**Pull request merged**